### PR TITLE
Move NaN/Inf detection to a separate utilities file

### LINKED ISSUE #6815: [trainer] Simplify Trainer dependencies by making TrainerTrainingTricksMixin a utils class
## 🚀 Feature

Training tricks doesn't need to be inherited from the core Trainer class. These methods could be utility functions that sit completely outside the Trainer class hierarchy. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/bb9ace43334ad50e3758d9cff08ad34216c7d4da/pytorch_lightning/trainer/training_tricks.py#L28

### Motivation

Simplify the Trainer inheritance hierarchy and create more reusable utilities

### Pitch

Deprecate the mixin and move the functions out to a utilities file for NaN detection. All we have to do is accept an nn.Module as input and pass the LightningModule to these functions from within the training loop.

1.
```
 def print_nan_gradients(model: nn.Module) -> None:
```
2.
```
def detect_nan_tensors(model: nn.Module) -> None:
```
we can separately check for the loss being NaN inside the training loop directly, since thats a specific call here: https://github.com/PyTorchLightning/pytorch-lightning/blob/bb9ace43334ad50e3758d9cff08ad34216c7d4da/pytorch_lightning/trainer/training_tricks.py#L43-L45

Deprecation path:
- Add the new utilities in a standalone file
- Make the mixin call the new utilities
- Update the training loop to directly call the new utilities
- Add a deprecation warning to the methods inside the mixin that mark them as deprecated

### Alternatives

Keep as is

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #6815

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
maybe add a todo on top of the TrainingTricksMixin class that it should be removed in v1.5?

File: pytorch_lightning/utilities/nan.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright The PyTorch Lightning team.
```
Comment (ananthsub) [inline pytorch_lightning/utilities/nan.py:1]: could be named `non_finite` to generalize NaN/Inf handling

File: pytorch_lightning/trainer/training_tricks.py [hunk 2]
```diff
@@ -19,35 +19,42 @@
...
+from pytorch_lightning.utilities.nan import detect_nan_parameters, print_nan_gradients
 
 EPSILON = 1e-6
 EPSILON_FP16 = 1e-5
 log = logging.getLogger(__name__)
 
 
 class TrainerTrainingTricksMixin(ABC):
+    """
+    TODO: This class will be deprecated in v1.5.
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_tricks.py:32]: Is this what you meant?

```suggestion
    TODO: Remove this class in v1.5.
```

File: pytorch_lightning/utilities/nan.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import logging
+
+import torch
+import torch.nn as nn
+
+log = logging.getLogger(__name__)
+
+
+def print_nan_gradients(model: nn.Module) -> None:
```
Comment (carmocca) [inline pytorch_lightning/utilities/nan.py:24]: What do you think of these improvements? (my code)

https://github.com/jpuigcerver/PyLaia/blob/master/laia/utils/checks.py

- Optional exception
- Include non finite percentages

Comment (tchaton) [reply pytorch_lightning/utilities/nan.py:24]: It could be useful for some users with unstable losses. Worth to consider.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -689,6 +690,12 @@ def training_step_and_backward(self, split_batch, batch_idx, opt_idx, optimizer,
 
         return result
 
+    def _check_nan(self, loss: torch.Tensor) -> None:
+        if not torch.isfinite(loss).all():
+            raise ValueError('The loss returned in `training_step` is nan or inf.')
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:695]:
```suggestion
            raise RuntimeError(f'The loss returned in `training_step` is {loss}.')
```

File: pytorch_lightning/trainer/training_tricks.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py:32]: rather make `#todo: ...` above the function, so IDE better index todos

File: pytorch_lightning/utilities/nan.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/utilities/nan.py:1]: rather `feasible_values`? 
the `nan` can be simply swapped with just a values e.g. `np.nan`

File: pytorch_lightning/utilities/nan.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/nan.py:24]: I think it'd be useful! Maybe we can add it in a follow up PR? this one is mainly moving code around for parity

Comment (carmocca) [reply pytorch_lightning/utilities/nan.py:24]: If you prefer, sure

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -764,7 +764,7 @@ def training_step(self, batch, batch_idx, optimizer_idx=None):
         terminate_on_nan=True,
     )
 
-    with pytest.raises(ValueError, match=r".*The loss returned in `training_step` is nan or inf.*"):
+    with pytest.raises(ValueError, match=r".*The loss returned in `training_step` is.*"):
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:767]: @Borda - raising a RuntimeError runs into the issues here: https://github.com/PyTorchLightning/pytorch-lightning/issues/6842

The trainer catches it so the exception is never raised

File: pytorch_lightning/utilities/nan.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/nan.py:1]: @ananthsub Are you making this change?  I'd say `finite_checks.py` sounds self-explanatory

Comment (ananthsub) [reply pytorch_lightning/utilities/nan.py]: renamed to `finite_checks`

**Review by carmocca — APPROVED** [review]
